### ISSUE TITLE
spaces in comprehensions

### ISSUE THREAD

**Issue by mistguy** [JuliaLang/julia#27400]
On both 0.7.0-alpha and 0.6.3:
```julia
julia> [1 +1 for i in 1:2]
ERROR: syntax: invalid comprehension syntax

julia> [1+1 for i in 1:2]
2-element Array{Int64,1}:
 2
 2

julia> [1+ 1 for i in 1:2]
2-element Array{Int64,1}:
 2
 2

julia> [1 + 1 for i in 1:2]
2-element Array{Int64,1}:
 2
 2
```

I think the intention of `[1 +1 for i in 1:2]` is clearly `1+1`, not `[1, +1]`, unless there are some other confusing situations?

**Comment by wuhuagumu** [comment]
Is it because, `i +j` will be resolved as `i positive j `? Maybe the reason is that list comprehension supports blank spaces, so it will happen only in lists? 

I noticed this one will have the same error reported:

    julia> [1 2 for i in 1:2]
    ERROR: ParseError("invalid comprehension syntax")
    Stacktrace:
     [1] #parse#236(::Bool, ::Bool, ::Function, ::String, ::Int64) at ./parse.jl:222
     [2] (::Base.#kw##parse)(::Array{Any,1}, ::Base.#parse, ::String, ::Int64) at./<missing>:0
     [3] #parse#237(::Bool, ::Function, ::String) at ./parse.jl:232
     [4] macro expansion at /home/liuxy/.julia/v0.6/Atom/src/repl.jl:118 [inlined]
     [5] anonymous at ./<missing>:?
     [6] macro expansion at ./REPL.jl:97 [inlined]
     [7] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
    
    julia> [[i +j for i in 1:2] for j in 1:2]
    ERROR: ParseError("invalid comprehension syntax")
    Stacktrace:
     [1] #parse#236(::Bool, ::Bool, ::Function, ::String, ::Int64) at ./parse.jl:222
     [2] (::Base.#kw##parse)(::Array{Any,1}, ::Base.#parse, ::String, ::Int64) at ./<missing>:0
     [3] #parse#237(::Bool, ::Function, ::String) at ./parse.jl:232
     [4] macro expansion at /home/liuxy/.julia/v0.6/Atom/src/repl.jl:118 [inlined]
     [5] anonymous at ./<missing>:?
     [6] macro expansion at ./REPL.jl:97 [inlined]
     [7] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73

**Comment by tomaklutfu** [comment]
Though intention is that so, I think it is ambiguous with infix `+` methods.

**Comment by garrison** [comment]
Another one I've noticed:
```julia
julia> [1 +
        x for x in 1:4]
ERROR: syntax: invalid comprehension syntax
```
See also #9494

**Comment by JeffBezanson** [comment]
The issue here is hcat syntax, where `[1 -2]` makes a 1x2 array. So it thinks `[1 +1 ...`is an hcat, at which point it's too late for it to be a comprehension. The case with the newline,
```
[1 +
 x for ...
```
seems like it could maybe be made to work.

As a footnote, since it's come up in the past, note that
```
[1
 + x for ...
```
is never going to work. Since we don't require `;` statement separators, we can't join expressions with operators on the next line.

**Comment by JeffBezanson** [comment]
The case with the line break was fixed by #20265. The other case I don't think can be fixed.
